Replace convoluted indexOfTheOnlyBit implementation with ctz# primop

Fuuzetsu opened the pull request:
Performance is the same or slightly better in microbenchmark. I
suspect it can perform better in the wild.

I was also interested in `revNat` but GHC doesn't expose an
appropriate primop, presumably because GCC doesn't. It could still be
faster on clang (`__builtin_bitreverse64` and friends).

Fuuzetsu commented:
I realised that there's `countTrailingZeroes` function in Data.Bits.FiniteBits typeclass, it's probably better to use that instead of GHC.Prim.

Fuuzetsu commented:
IIRC I ran the benchmarks at the time and they were pretty much the same (so that's a good thing). I don't have the numbers at hand, I can try to run them again after changing to countTrailingZeroes and post the results.

I see that CI failed due to old GHC versions which didn't have the primop: are these versions we care about and I should add CPP?

w.r.t. reversing bits, let's forget about that for now. I made a GHC ticket if you're interested: https://ghc.haskell.org/trac/ghc/ticket/16164

treeowl commented:
Yes, I think we still have to care about GHC 7.8 for another major version. Let's use CPP for now.

Fuuzetsu commented:
I added some CPP, not using the primop directly and rebased. The benchmarks are above. I don't think there's any real impact on performance that I can accurately measure on my fairly noisy machine but please look through yourself.

current master
```
Preprocessing library for containers-0.6.0.1..
Building library for containers-0.6.0.1..
Preprocessing benchmark 'intset-benchmarks' for containers-0.6.0.1..
Building benchmark 'intset-benchmarks' for containers-0.6.0.1..
Running 1 benchmarks...
Benchmark intset-benchmarks: RUNNING...
benchmarking member
time                 62.90 μs   (62.39 μs .. 63.35 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 62.71 μs   (62.32 μs .. 63.24 μs)
std dev              1.483 μs   (1.149 μs .. 2.135 μs)
variance introduced by outliers: 20% (moderately inflated)

benchmarking insert
time                 65.34 μs   (65.00 μs .. 65.74 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 65.63 μs   (65.30 μs .. 66.07 μs)
std dev              1.296 μs   (962.6 ns .. 1.950 μs)
variance introduced by outliers: 15% (moderately inflated)

benchmarking map
time                 99.63 μs   (98.32 μs .. 101.2 μs)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 98.89 μs   (98.14 μs .. 100.1 μs)
std dev              3.015 μs   (2.079 μs .. 4.848 μs)
variance introduced by outliers: 29% (moderately inflated)

benchmarking filter
time                 17.89 μs   (17.61 μs .. 18.12 μs)
                     0.999 R²   (0.999 R² .. 0.999 R²)
mean                 17.62 μs   (17.49 μs .. 17.77 μs)
std dev              516.1 ns   (438.3 ns .. 641.0 ns)
variance introduced by outliers: 32% (moderately inflated)

benchmarking partition
time                 19.45 μs   (19.15 μs .. 19.75 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 19.23 μs   (19.12 μs .. 19.41 μs)
std dev              492.7 ns   (363.3 ns .. 631.5 ns)
variance introduced by outliers: 27% (moderately inflated)

benchmarking fold
time                 146.6 ns   (145.6 ns .. 147.6 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 145.5 ns   (144.9 ns .. 146.2 ns)
std dev              2.160 ns   (1.700 ns .. 2.886 ns)
variance introduced by outliers: 17% (moderately inflated)

benchmarking delete
time                 90.17 μs   (89.68 μs .. 90.76 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 90.36 μs   (89.90 μs .. 91.21 μs)
std dev              2.109 μs   (1.466 μs .. 2.944 μs)
variance introduced by outliers: 19% (moderately inflated)

benchmarking findMin
time                 14.55 ns   (14.45 ns .. 14.68 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.49 ns   (14.45 ns .. 14.55 ns)
std dev              143.4 ps   (98.51 ps .. 229.1 ps)

benchmarking findMax
time                 7.984 ns   (7.943 ns .. 8.041 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 8.155 ns   (8.097 ns .. 8.214 ns)
std dev              202.2 ps   (184.0 ps .. 231.7 ps)
variance introduced by outliers: 41% (moderately inflated)

benchmarking deleteMin
time                 80.77 ns   (80.28 ns .. 81.18 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 80.90 ns   (80.46 ns .. 81.57 ns)
std dev              1.766 ns   (1.415 ns .. 2.141 ns)
variance introduced by outliers: 31% (moderately inflated)

benchmarking deleteMax
time                 23.31 ns   (23.17 ns .. 23.46 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 23.02 ns   (22.91 ns .. 23.20 ns)
std dev              443.7 ps   (324.8 ps .. 692.1 ps)
variance introduced by outliers: 28% (moderately inflated)

benchmarking unions
time                 644.9 ns   (639.2 ns .. 651.4 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 642.5 ns   (638.1 ns .. 648.6 ns)
std dev              16.52 ns   (12.72 ns .. 21.78 ns)
variance introduced by outliers: 35% (moderately inflated)

benchmarking union
time                 637.9 ns   (633.0 ns .. 642.1 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 636.8 ns   (634.8 ns .. 639.3 ns)
std dev              7.342 ns   (6.155 ns .. 8.978 ns)

benchmarking difference
time                 736.8 ns   (728.0 ns .. 747.1 ns)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 731.0 ns   (725.4 ns .. 740.6 ns)
std dev              25.20 ns   (16.88 ns .. 41.75 ns)
variance introduced by outliers: 49% (moderately inflated)

benchmarking intersection
time                 703.2 ns   (696.4 ns .. 712.3 ns)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 698.2 ns   (695.5 ns .. 703.7 ns)
std dev              12.92 ns   (8.059 ns .. 22.23 ns)
variance introduced by outliers: 21% (moderately inflated)

benchmarking fromList
time                 65.43 μs   (64.88 μs .. 65.97 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 65.54 μs   (65.19 μs .. 65.93 μs)
std dev              1.276 μs   (1.062 μs .. 1.609 μs)
variance introduced by outliers: 15% (moderately inflated)

benchmarking fromAscList
time                 33.32 μs   (33.17 μs .. 33.50 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 33.48 μs   (33.30 μs .. 33.77 μs)
std dev              736.2 ns   (493.5 ns .. 1.058 μs)
variance introduced by outliers: 20% (moderately inflated)

benchmarking fromDistinctAscList
time                 12.13 μs   (12.05 μs .. 12.22 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 12.22 μs   (12.11 μs .. 12.36 μs)
std dev              387.0 ns   (318.6 ns .. 453.0 ns)
variance introduced by outliers: 37% (moderately inflated)

benchmarking disjoint:false
time                 38.18 ns   (37.97 ns .. 38.44 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 38.92 ns   (38.62 ns .. 39.27 ns)
std dev              1.087 ns   (941.9 ps .. 1.257 ns)
variance introduced by outliers: 44% (moderately inflated)

benchmarking disjoint:true
time                 404.2 ns   (402.4 ns .. 406.8 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 407.5 ns   (405.1 ns .. 411.7 ns)
std dev              11.27 ns   (7.557 ns .. 19.01 ns)
variance introduced by outliers: 39% (moderately inflated)

benchmarking null.intersection:false
time                 701.9 ns   (696.2 ns .. 709.7 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 700.7 ns   (698.7 ns .. 703.7 ns)
std dev              8.150 ns   (5.561 ns .. 13.55 ns)

benchmarking null.intersection:true
time                 423.3 ns   (420.2 ns .. 427.1 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 423.8 ns   (421.2 ns .. 427.2 ns)
std dev              9.861 ns   (8.194 ns .. 12.42 ns)
variance introduced by outliers: 31% (moderately inflated)

Benchmark intset-benchmarks: FINISH
```

after my change
```
Preprocessing library for containers-0.6.0.1..
Building library for containers-0.6.0.1..
Preprocessing benchmark 'intset-benchmarks' for containers-0.6.0.1..
Building benchmark 'intset-benchmarks' for containers-0.6.0.1..
Running 1 benchmarks...
Benchmark intset-benchmarks: RUNNING...
benchmarking member
time                 63.33 μs   (62.71 μs .. 64.19 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 63.60 μs   (63.18 μs .. 64.28 μs)
std dev              1.883 μs   (1.318 μs .. 2.927 μs)
variance introduced by outliers: 29% (moderately inflated)

benchmarking insert
time                 65.84 μs   (65.05 μs .. 66.72 μs)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 65.54 μs   (64.96 μs .. 66.35 μs)
std dev              2.318 μs   (1.755 μs .. 2.932 μs)
variance introduced by outliers: 37% (moderately inflated)

benchmarking map
time                 97.87 μs   (96.29 μs .. 99.65 μs)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 96.39 μs   (95.87 μs .. 97.32 μs)
std dev              2.430 μs   (1.752 μs .. 3.328 μs)
variance introduced by outliers: 22% (moderately inflated)

benchmarking filter
time                 17.09 μs   (17.00 μs .. 17.19 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 17.05 μs   (16.99 μs .. 17.14 μs)
std dev              250.0 ns   (194.1 ns .. 326.2 ns)
variance introduced by outliers: 11% (moderately inflated)

benchmarking partition
time                 18.47 μs   (18.20 μs .. 18.87 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 18.22 μs   (18.16 μs .. 18.43 μs)
std dev              335.3 ns   (159.8 ns .. 639.6 ns)
variance introduced by outliers: 16% (moderately inflated)

benchmarking fold
time                 146.0 ns   (144.2 ns .. 147.6 ns)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 147.8 ns   (145.0 ns .. 154.1 ns)
std dev              14.45 ns   (7.025 ns .. 27.82 ns)
variance introduced by outliers: 90% (severely inflated)

benchmarking delete
time                 91.76 μs   (90.79 μs .. 92.79 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 91.32 μs   (90.44 μs .. 92.20 μs)
std dev              2.888 μs   (2.471 μs .. 3.435 μs)
variance introduced by outliers: 31% (moderately inflated)

benchmarking findMin
time                 15.05 ns   (14.94 ns .. 15.21 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 14.96 ns   (14.90 ns .. 15.07 ns)
std dev              272.1 ps   (123.8 ps .. 506.4 ps)
variance introduced by outliers: 26% (moderately inflated)

benchmarking findMax
time                 7.701 ns   (7.672 ns .. 7.741 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 7.714 ns   (7.683 ns .. 7.759 ns)
std dev              125.5 ps   (77.44 ps .. 203.1 ps)
variance introduced by outliers: 23% (moderately inflated)

benchmarking deleteMin
time                 76.61 ns   (76.02 ns .. 77.59 ns)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 76.88 ns   (76.43 ns .. 77.82 ns)
std dev              2.182 ns   (1.036 ns .. 4.176 ns)
variance introduced by outliers: 44% (moderately inflated)

benchmarking deleteMax
time                 20.16 ns   (19.67 ns .. 20.70 ns)
                     0.997 R²   (0.996 R² .. 1.000 R²)
mean                 19.89 ns   (19.73 ns .. 20.13 ns)
std dev              674.3 ps   (442.1 ps .. 1.019 ns)
variance introduced by outliers: 55% (severely inflated)

benchmarking unions
time                 687.8 ns   (681.3 ns .. 694.7 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 693.4 ns   (687.2 ns .. 701.6 ns)
std dev              24.36 ns   (19.25 ns .. 30.23 ns)
variance introduced by outliers: 50% (severely inflated)

benchmarking union
time                 678.8 ns   (670.3 ns .. 688.3 ns)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 672.8 ns   (669.7 ns .. 679.6 ns)
std dev              14.98 ns   (8.864 ns .. 25.68 ns)
variance introduced by outliers: 28% (moderately inflated)

benchmarking difference
time                 690.9 ns   (681.9 ns .. 701.6 ns)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 684.1 ns   (680.0 ns .. 694.8 ns)
std dev              19.77 ns   (10.17 ns .. 41.40 ns)
variance introduced by outliers: 40% (moderately inflated)

benchmarking intersection
time                 737.4 ns   (733.1 ns .. 741.8 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 740.0 ns   (736.0 ns .. 745.5 ns)
std dev              14.90 ns   (11.18 ns .. 19.38 ns)
variance introduced by outliers: 24% (moderately inflated)

benchmarking fromList
time                 64.46 μs   (64.05 μs .. 64.87 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 65.02 μs   (64.54 μs .. 65.81 μs)
std dev              2.023 μs   (1.488 μs .. 2.671 μs)
variance introduced by outliers: 31% (moderately inflated)

benchmarking fromAscList
time                 32.85 μs   (32.69 μs .. 33.02 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 32.77 μs   (32.68 μs .. 32.88 μs)
std dev              333.7 ns   (291.9 ns .. 389.0 ns)

benchmarking fromDistinctAscList
time                 12.01 μs   (11.94 μs .. 12.10 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 12.06 μs   (11.99 μs .. 12.16 μs)
std dev              283.9 ns   (217.8 ns .. 364.0 ns)
variance introduced by outliers: 24% (moderately inflated)

benchmarking disjoint:false
time                 33.00 ns   (32.56 ns .. 33.42 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 32.81 ns   (32.51 ns .. 33.71 ns)
std dev              1.600 ns   (696.3 ps .. 3.298 ns)
variance introduced by outliers: 71% (severely inflated)

benchmarking disjoint:true
time                 381.5 ns   (379.6 ns .. 384.9 ns)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 383.8 ns   (382.2 ns .. 386.2 ns)
std dev              6.428 ns   (4.497 ns .. 9.753 ns)
variance introduced by outliers: 19% (moderately inflated)

benchmarking null.intersection:false
time                 728.0 ns   (724.8 ns .. 731.9 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 726.2 ns   (724.2 ns .. 728.9 ns)
std dev              7.760 ns   (5.888 ns .. 10.72 ns)

benchmarking null.intersection:true
time                 393.8 ns   (392.5 ns .. 395.8 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 398.5 ns   (396.1 ns .. 401.9 ns)
std dev              9.770 ns   (6.413 ns .. 15.23 ns)
variance introduced by outliers: 33% (moderately inflated)

Benchmark intset-benchmarks: FINISH
```

Fuuzetsu commented:
This actually looks like duplicate of #379...

ethercrow commented:
If someone is curious enough, please try benchmarking with GHC HEAD and `-mbmi2`.

ethercrow commented:
This one also might benefit from https://gitlab.haskell.org/ghc/ghc/merge_requests/27

treeowl commented:
Can you rebase this? I would very much like to get rid of that convoluted mess....

treeowl commented:
Actually, #379 looks like it goes a bit further. Let's do that instead.
